### PR TITLE
Apply assorted ruff rules

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
Subject: apply new ruff rules

### Feature or Bugfix
- Refactoring

### CODE REVIEW

File: sphinx/config.py [hunk 1]
```diff
@@ -92,7 +92,7 @@ def match(self, value: str | list | tuple) -> bool:
 
 
 class _Opt:
-    __slots__ = 'default', 'rebuild', 'valid_types', 'description'
+    __slots__ = 'default', 'description', 'rebuild', 'valid_types'
```
Comment (AA-Turner) [inline sphinx/config.py:95]: Why this change? The slots are in definition order

File: tests/test_util/typing_test_data.py [hunk 2]
```diff
@@ -121,7 +121,7 @@ def f25(a, b, /):
     pass
 
 
-def f26(x: Literal[1, 2, 3] = 1, y: Union[Literal['a'], Literal['b']] = 'a') -> None:
+def f26(x: Literal[1, 2, 3] = 1, y: Literal['a', 'b'] = 'a') -> None:
```
Comment (AA-Turner) [inline tests/test_util/typing_test_data.py]: Typing test data should continue to test unions of literals

**Comment by AA-Turner** [comment]
PRs which apply multiple rules at once make history harder to decipher, perhaps we could split to one PR per rule, or per group of rules.

A

File: sphinx/config.py (hunk 1 above)
Comment (DimitriPapadopoulos) [reply sphinx/config.py:95]: This is indeed debatable. The rationale for [RUF023](https://docs.astral.sh/ruff/rules/unsorted-dunder-slots/) is:
> Checks for `__slots__` definitions that are not ordered according to a [natural sort](https://en.wikipedia.org/wiki/Natural_sort_order).

> Consistency is good. Use a common convention for this special variable to make your code more readable and idiomatic.

**Comment by DimitriPapadopoulos** [comment]
Alternatively, merge without squashing. The PR is already split into multiple commits.

File: tests/test_util/typing_test_data.py (hunk 2 above)
Comment (DimitriPapadopoulos) [reply tests/test_util/typing_test_data.py]: 👍 Fixed.

**Comment by AA-Turner** [comment]
Merge commits are disabled for this repo -- there's no option to select it. This is to preserve linear history.
